### PR TITLE
Code Optimization

### PR DESCRIPTION
**Pull request by kompilorb**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Comments from other PR : https://github.com/PrestaShop/PrestaShop/pull/7816
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/pull/7816
| How to test?  | Create in BO a voucher (the problems is for all type or rules). After create a customers, and create you cart, if the first product does not in condition for rules, if after you add product corresponding to a rules, voucher must can be applicable.

### CODE REVIEW

**Comment by kompilorb** [comment]
@eternoendless

File: classes/CartRule.php [hunk 1]
```diff
@@ -38,7 +38,9 @@ class CartRuleCore extends ObjectModel
     const FILTER_ACTION_ALL_NOCAP = 5;
     const BO_ORDER_CODE_PREFIX = 'BO_ORDER_';
 
-    /* This variable controls that a free gift is offered only once, even when multi-shippping is activated 
+    /* @var array $only_one_gift
```
Comment (eternoendless) [inline classes/CartRule.php]: Don't forget that in order for a comment to be a valid PHPDoc it must start with `/**` and not `/*` (see [the proposed standard](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#3-definitions)).

Also, keep in mind that when documenting class properties, you don't need to write the property name again.

In addition, I think that for properties it is common practice to write the description before the `@var` statement.

Here's my proposal:

```php
    /**
     * This variable controls that a free gift is offered only once, even when multi-shippping is activated
     * and the same product is delivered in both addresses
     *
     * @var array
     */
```

File: classes/CartRule.php [hunk 2]
```diff
@@ -1575,34 +1586,31 @@ public static function getCartsRuleByCode($name, $id_lang, $extended = false)
...
         } else {
-            $productList = $arrayProduct;
+            $productsList = $arrayProduct;
         }
 
-        foreach ($productList as $product) {
-            if (in_array($product, $arrayRules) && isset($arrayProduct[$array])) {
-                $return[] = $arrayProduct[$array];
+        foreach ($productsList as $k => $product) {
+            if (in_array($products, $eligibleProducts)) {
```
Comment (eternoendless) [inline classes/CartRule.php]: I think you mean `$product` instead of `$products` here, don't you?

File: classes/CartRule.php [hunk 3]
```diff
@@ -285,7 +289,8 @@ public static function getIdByCode($code)
             return false;
         }
         return Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue(
-                'SELECT `id_cart_rule` FROM `' . _DB_PREFIX_ . 'cart_rule` WHERE `code` = \'' . pSQL($code) . '\'');
+                'SELECT `id_cart_rule` FROM `' . _DB_PREFIX_ . 'cart_rule` WHERE `code` = \'' . pSQL($code) . '\''
+            );
```
Comment (eternoendless) [inline classes/CartRule.php]: Whoops, I think this block needs to be indented one step to the left.

File: classes/CartRule.php [hunk 4]
```diff
@@ -1575,34 +1586,31 @@ public static function getCartsRuleByCode($name, $id_lang, $extended = false)
...
+     * @param array $products List of Products from the cart,
+     * @param array $eligibleProducts List of Product eligible for rules,
+     * @param string $ruleType name of the rule,
      *
-     * @return int 0 = same
-     *             1 = different
+     * @return array Product selected who are eligible
      */
-    public function cartRulesCompare($arrayProduct, $arrayRules, $type)
+    protected function filterProducts($products, $eligibleProducts, $ruleType)
```
Comment (eternoendless) [inline classes/CartRule.php]: Thank you for changing the method signature, but it looks like you forgot to change the method calls!

Also, please remember to make them instance calls if you don't need them to be static.

**Review by eternoendless — CHANGES_REQUESTED** [review]
Thanks for this PR! There are still two or three small things to fix.

File: classes/CartRule.php (hunk 2 above)
Comment (Shudrum) [inline classes/CartRule.php]: +1
How can it just works? 🤕

**Review by eternoendless — APPROVED** [review]
Thanks!

File: classes/CartRule.php (hunk 4 above)
Comment (maximebiloe) [inline classes/CartRule.php]: We can't simply delete a public method like this. Please, deprecate it first for `1.7.2`, then we'll be able to delete it.

**Review by maximebiloe — CHANGES_REQUESTED** [review]

File: classes/CartRule.php (hunk 4 above)
Comment (Shudrum) [inline classes/CartRule.php]: This method added on a PR just before. No problem, this is just to avoid this kind of problem!

Comment (Shudrum) [inline classes/CartRule.php]: +was -_-

Comment (maximebiloe) [inline classes/CartRule.php]: Ok :wink:

**Review by maximebiloe — APPROVED** [review]

File: classes/CartRule.php [hunk 5]
```diff
@@ -1313,7 +1319,13 @@ protected function getCartRuleCombinations($offset = null, $limit = null, $searc
      * @return array|bool Array with DB rows of requested type
      * @throws PrestaShopDatabaseException
      */
-    public function getAssociatedRestrictions($type, $active_only, $i18n, $offset = null, $limit = null, $search_cart_rule_name = '')
+    public function getAssociatedRestrictions(
+        $type,
+        $active_only,
+        $i18n, $offset = null,
```
Comment (Quetzacoalt91) [inline classes/CartRule.php]: If you wanted to write param perline, maybe you would like to separate `$i18n` from `$offset` ?

File: classes/CartRule.php [hunk 6]
```diff
@@ -1575,34 +1587,31 @@ public static function getCartsRuleByCode($name, $id_lang, $extended = false)
...
-     *             1 = different
+     * @return array Product selected who are eligible
      */
-    public function cartRulesCompare($arrayProduct, $arrayRules, $type)
+    protected function filterProducts($products, $eligibleProducts, $ruleType)
     {
         //If the two same array, no verification todo.
-        if ($arrayProduct == $arrayRules) {
+        if ($products === $eligibleProducts) {
             return $arrayProduct;
```
Comment (Quetzacoalt91) [inline classes/CartRule.php]: Now you renamed the parameter, `$arrayProduct` is undefined

File: classes/CartRule.php [hunk 7]
```diff
@@ -1575,34 +1587,31 @@ public static function getCartsRuleByCode($name, $id_lang, $extended = false)
...
-        $array = 0;
         // Attribute id is not important for this filter in the global list
         // so the ids are replaced by 0
-        if (in_array($type, array('products', 'categories', 'manufacturers', 'suppliers'))) {
-            $productList = explode(':', preg_replace("#\-[0-9]+#", "-0", implode(':', $arrayProduct)));
+        if (in_array($ruleType, array('products', 'categories', 'manufacturers', 'suppliers'))) {
+            $productsList = explode(':', preg_replace("#\-[0-9]+#", "-0", implode(':', $arrayProduct)));
         } else {
-            $productList = $arrayProduct;
+            $productsList = $arrayProduct;
```
Comment (Quetzacoalt91) [inline classes/CartRule.php]: `$arrayProduct` is undefined

File: classes/CartRule.php [hunk 8]
```diff
@@ -1575,34 +1587,31 @@ public static function getCartsRuleByCode($name, $id_lang, $extended = false)
...
         }
 
-        foreach ($productList as $product) {
-            if (in_array($product, $arrayRules) && isset($arrayProduct[$array])) {
-                $return[] = $arrayProduct[$array];
+        foreach ($productsList as $k => $product) {
+            if (in_array($product, $eligibleProducts)) {
+                $return[] = $arrayProduct[$k];
```
Comment (Quetzacoalt91) [inline classes/CartRule.php]: `$arrayProduct` is undefined

**Review by Quetzacoalt91 — CHANGES_REQUESTED** [review]
Sorry dude, the previous reviews brought you new changes to apply :stuck_out_tongue:

**Review by Quetzacoalt91 — APPROVED** [review]

**Pull request merged**